Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +25,71 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return ampHours_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    return (current_ < 0);
+}
+
+int BatteryStateOfChargeService::getRemainingHours() const
+{
+    return remainingHours_;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    return timeTillChargeOrDepletion_;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
-    // Hint: There are many different ways that the totalAmpHoursUsed can be updated
-    // i.e: Taking a running average of your data values, using most recent data points, etc.
+    QTime timeOld = time_;
+    double currentOld = current_;
+
+    current_ = batteryData.current;
+    time_ = batteryData.time;
+
+    if(!timeOld.isNull())
+    {
+        double avgCurrent = (currentOld + current_) / 2;
+        ampHours_ += (avgCurrent * timeOld.secsTo(time_) / SECONDS_TO_HOURS);
+    }
+
+    //calculate remaining time for charge or depletion
+    double totalHours;
+
+    if(isCharging())
+    {
+        totalHours = qAbs(totalAmpHoursUsed() / current_);
+    }
+    else
+    {
+        totalHours = (BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed()) / current_;
+    }
+
+    int milliseconds = totalHours * MILLISECONDS_IN_HOUR;
+
+    int hours = floor(totalHours);
+    milliseconds -= hours * MILLISECONDS_IN_HOUR;
+
+    int minutes = floor(milliseconds / MILLISECONDS_IN_MINUTE);
+    milliseconds -= minutes * MILLISECONDS_IN_MINUTE;
+
+    int seconds = floor(milliseconds / MILLISECONDS_IN_SECOND);
+    milliseconds -= seconds * MILLISECONDS_IN_SECOND;
+
+    if((hours - HOURS_IN_A_DAY) >= 0) // if 24 hours or more
+    {
+        remainingHours_ = hours - (HOURS_IN_A_DAY - 1); // calculate the amount of hours needed to take out of "hours" as "hours" can only hold up to 23
```

Comment:
This line is getting too long, let's put the comment the line above. No strict rule on line length, but a good rule of thumb that if it can't fit when you do split screen on your editor, it's too long. 

Also I think we can condense this comment to just: `QTime can only hold up to 23 hours` 
Unless the code isn't straightforward enough to just read, most comments should be supplementary - explaining the intention behind the change, instead of what the change does.